internal/pkg/service/stream/storage/node/readernode/sliceupload/operator.go

```diff
@@ -0,0 +1,363 @@
+				return slice.SliceKey
+			},
+			func(_ string, slice model.Slice, rawValue *op.KeyValue, oldValue **uploadedSliceData) *uploadedSliceData {
+				out := &uploadedSliceData{
+					filename: slice.String(),
+				}
+
+				// Keep the same lock, to prevent parallel processing of the same slice.
+				// No modification from another code is expected, but just to be sure.
+				if oldValue != nil {
+					out.Lock = (*oldValue).Lock
+				} else {
+					out.Lock = &sync.Mutex{}
+				}
+
+				return out
+			},
+		).
+			WithFilter(func(event etcdop.WatchEvent[model.Slice]) bool {
+				return event.Value.State == model.SliceUploaded
+			}).
+			BuildMirror()
+		if err = <-o.uploadedSlices.StartMirroring(ctx, wg, o.logger); err != nil {
+			return err
+		}
+	}
+	// Start conditions check ticker
+	{
+		wg.Add(1)
+		ticker := d.Clock().Ticker(o.config.CheckInterval.Duration())
+
+		go func() {
+			defer wg.Done()
+			defer ticker.Stop()
+
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case <-ticker.C:
+					o.checkSlices(ctx, wg)
+				}
+			}
+		}()
+	}
+
+	return nil
+}
+
+func (o *operator) checkSlices(ctx context.Context, wg *sync.WaitGroup) {
+	o.logger.Debugf(ctx, "checking slices upload conditions")
+
+	o.slices.ForEach(func(_ model.SliceKey, data *sliceData) (stop bool) {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			o.checkSlice(ctx, data)
+		}()
+		return false
+	})
+}
+
+func (o *operator) checkSlice(ctx context.Context, data *sliceData) {
+	// Prevent multiple checks of the same slice
+	if !data.Lock.TryLock() {
+		return
+	}
+	defer data.Lock.Unlock()
+
+	// No plugin existing for given slice to be uploaded
+	err := o.plugins.ImporterFor(data.Slice.StagingStorage.Provider)
+	if err != nil {
+		o.logger.Errorf(ctx, "importer for provider: %v does not exists", data.Slice.StagingStorage.Provider)
+		return
+	}
+
+	// Slice uploading is already in progress
+	if data.Uploading {
+		return
+	}
+
+	// Slice has been modified by some previous check, but we haven't received an updated version from etcd yet
+	if data.Processed {
+		return
+	}
+
+	if !data.Slice.Retryable.Allowed(o.clock.Now()) {
+		return
+	}
+
+	volume, err := o.volumes.Collection().Volume(data.Slice.SliceKey.VolumeID)
+	if err != nil {
+		o.logger.Errorf(ctx, "unable to upload slice. No volume found for key: %v", data.Slice.SliceKey.VolumeID)
+		return
+	}
+
+	switch data.Slice.State {
+	case model.SliceUploading:
+		o.uploadSlice(ctx, volume, data)
+	default:
+		// nop
+	}
+}
+
+func (o *operator) uploadSlice(ctx context.Context, volume *diskreader.Volume, data *sliceData) {
+	// Empty slice does not need to be uploaded to staging. Just mark as uploaded
+	if data.Slice.LocalStorage.IsEmpty {
+		// Update the entity, the ctx may be cancelled
+		dbCtx, dbCancel := context.WithTimeout(context.WithoutCancel(ctx), 30*time.Second)
+		defer dbCancel()
+
+		err := o.storage.Slice().SwitchToUploaded(data.Slice.SliceKey, o.clock.Now()).Do(dbCtx).Err()
```

Comment:
Empty slice is switched to `uploaded` right away as there is no data to be move to staging storage.

Reply:
Ok, but `slice upload` operator must set the `IsEmpty` flag first, by default is always `false`.